src/nrniv/partrans.cpp

```diff
@@ -1071,211 +1066,134 @@ extern size_t nrnbbcore_gap_write(const char* path, int* group_ids);
 
-  // type
   assert(target_pntlist_ && target_pntlist_->count() == targets_->count());
-  int type = target_pntlist_->item(0)->prop->type;
-  // all the same type
-  for (int i=0; i < targets_->count(); ++i) {
-    assert(type == target_pntlist_->item(i)->prop->type);
-  }
-
-  int ix_vpre = target_parray_index_->item(0);
 
   // space for the info
-  gi = new BBCoreGapInfo[nrn_nthread];
-  for (int tid = 0; tid < nrn_nthread; ++tid) {
-    BBCoreGapInfo& g = gi[tid];
-    int n = transfer_thread_data_[tid].cnt;
-    g.sid_src = new int[n]; // but perhaps fewer sources than targets
-    g.v_index = new int[n]; // fewer sources than targets
-    g.sid_target = new int[n];
-    g.vpre_index = new int[n];
-    g.ntar = 0; // will end up as gap_ml[tid].nodecount
-    g.nsrc = 0; // will end up as number of distinct gap_ml[tid].nodeindices
-
-    g.type = type;
-    g.ix_vpre = ix_vpre;
-  }
-
-  // memb_lists of the threads.
-  for (int tid=0; tid < nrn_nthread; ++tid) {
-    NrnThread& nt = nrn_threads[tid];
-    for (NrnThreadMembList* tml = nt.tml; tml; tml = tml->next) {
-      if (tml->index == type) {
-        gi[tid].gap_ml = tml->ml;
-      }
-    }
-  }
-
-  // gap info for targets, segregate into threads
-  for (int i=0; i < targets_->count(); ++i) {
-    int sid = sgid2targets_->item(i);
-    NrnThread* nt = (NrnThread*)target_pntlist_->item(i)->_vnt;
-    int tid = nt ? nt->id : 0;
-    Memb_list* ml = gi[tid].gap_ml;
-    int ix = targets_->item(i) - ml->data[0];
-
-    BBCoreGapInfo& g = gi[tid];
-    g.sid_target[g.ntar] = sid;
-    g.vpre_index[g.ntar] = ix;
-    g.ntar += 1;
-  }
-
-  // gap info for sources, segregate into threads.
-  for (int i=0; i < sgids_->count(); ++i) {
-    int sid = sgids_->item(i);
-    Node* nd = visources_->item(i);
-    assert(nd->extnode == NULL);
-    int tid = nd->_nt ? nd->_nt->id : 0;
-    int ix = nd->_v - nrn_threads[tid]._actual_v;
-    assert(ix >= 0 && ix < nrn_threads[tid].end);
-    BBCoreGapInfo& g = gi[tid];
-    g.sid_src[g.nsrc] = sid;
-    g.v_index[g.nsrc] = ix;
-    g.nsrc += 1;
-  }
-
-  // Sort each threads sources and targets so that the indices are
-  // monotonically increasing. (In fact the target indices will then be
-  // in memb_list order and that array will not have to be transferred.)
-  for (int tid=0; tid < nrn_nthread; ++tid) {
-    BBCoreGapInfo& g = gi[tid];
-    sidsort(g.sid_src, g.nsrc, g.v_index);
-    sidsort(g.sid_target, g.ntar, g.vpre_index);
+  if (gi) {
+    delete [] gi;
   }
-}
+  gi = new SetupTransferInfo[cn_nthread];
 
-static bool mycompar(const std::pair<int, int>& a, const std::pair<int, int>& b) {
-  return a.second < b.second;
-}
+  // info for targets, segregate into threads
+  if (targets_) {
+    for (int i=0; i < targets_->count(); ++i) {
+      sgid_t sid = sgid2targets_->item(i);
+      Point_process* pp = target_pntlist_->item(i);
+      NrnThread* nt = (NrnThread*)pp->_vnt;
+      int tid = nt ? nt->id : 0;
```

Comment:
curious - is it possible to have nt null here? because we dereference it on line 1159

Reply:
Great catch! The dereference should be ```Memb_list& ml = *(nrn_threads[tid]._ml_list[type]);``` But, actually, I don't really recall if _vnt is NULL when threads are not explicitly mentioned in Hoc or is set to nrn_threads. I did not test with an absent
```pc.nthread(1)```